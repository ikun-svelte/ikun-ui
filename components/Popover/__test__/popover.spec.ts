import { tick } from 'svelte';
import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';
import KPopoverPlacement from './popover.placement.svelte';
import KPopoverTrigger from './popover.trigger.svelte';
import KPopoverSlots from './popover.slots.svelte';
import KPopoverDisabled from './popover.disabled.svelte';
let host: HTMLElement;

const initHost = () => {
	host = document.createElement('div');
	host.setAttribute('id', 'host');
	document.body.appendChild(host);
};
beforeEach(() => {
	initHost();
	vi.useFakeTimers();
});
afterEach(() => {
	host.remove();
	vi.useRealTimers();
});

describe('Test: KPopover', () => {
	test('props: placement', async () => {
		const instance = new KPopoverPlacement({
			target: host
		});
		expect(instance).toBeTruthy();
		await tick();
		const Elm = host.children[0].getElementsByTagName('div')[0];
		Elm.dispatchEvent(
			new MouseEvent('mouseenter', {
				cancelable: true
			})
		);
		vi.runAllTimers();
		await tick();
		expect(host.innerHTML.includes('有美一人，清扬婉兮')).toBeTruthy();
		expect(host.innerHTML.includes('data-popper-arrow-bottom')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: disabled', async () => {
		const instance = new KPopoverDisabled({
			target: host
		});
		expect(instance).toBeTruthy();
		await tick();
		const Elm = host.children[0].getElementsByTagName('div')[0];
		Elm.dispatchEvent(
			new MouseEvent('mouseenter', {
				cancelable: true
			})
		);
		vi.runAllTimers();
		await tick();
		expect(host.innerHTML.includes('有美一人，清扬婉兮')).not.toBeTruthy();
		expect(host.innerHTML.includes('data-popper-arrow-bottom')).not.toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: trigger', async () => {
		const instance = new KPopoverTrigger({
			target: host
		});
		expect(instance).toBeTruthy();
		await tick();
		const Elm = host.children[0].getElementsByTagName('div')[0];
		Elm.click();
		vi.runAllTimers();
		await tick();
		expect(host.innerHTML.includes('有美一人，清扬婉兮')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('slots: triggerEl & contentEl', async () => {
		const instance = new KPopoverSlots({
			target: host
		});
		expect(instance).toBeTruthy();
		await tick();
		const Elm = host.children[0].getElementsByTagName('div')[0];
		Elm.dispatchEvent(
			new MouseEvent('mouseenter', {
				cancelable: true
			})
		);
		vi.runAllTimers();
		await tick();
		expect(
			host.innerHTML.includes('triggerEl') && host.innerHTML.includes('contentEl')
		).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});
});
