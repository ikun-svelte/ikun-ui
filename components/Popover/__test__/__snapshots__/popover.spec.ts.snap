// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test: KPopover > props: disabled 1`] = `"<div><div aria-hidden=\\"true\\"><div slot=\\"triggerEl\\" cls=\\"mx-2\\">top</div></div> </div>"`;

exports[`Test: KPopover > props: placement 1`] = `"<div><div aria-hidden=\\"true\\"><div slot=\\"triggerEl\\" cls=\\"mx-2\\">top</div></div> <div class=\\"k-popover--base k-popover--base__bottom k-popover--base__bottom__dark k-popover--base__dark svelte-1dixidq\\" data-popper-placement=\\"\\" aria-hidden=\\"true\\" style=\\"position: fixed; left: 0px; top: 0px; margin: 0px; animation: __svelte_185617033_0 200ms linear 0ms 1 both;\\"><div slot=\\"contentEl\\" class=\\"flex flex-col\\"><p class=\\"!my-2\\">有美一人，清扬婉兮</p></div> <div k-popover-arrow=\\"true\\" data-popper-arrow-bottom=\\"\\" data-popper-arrow-top=\\"\\" data-popper-arrow-right=\\"\\" data-popper-arrow-left=\\"\\" class=\\"svelte-1dixidq\\"></div></div></div>"`;

exports[`Test: KPopover > props: trigger 1`] = `"<div><div aria-hidden=\\"true\\"><div slot=\\"triggerEl\\" cls=\\"mx-2\\">top</div></div> <div class=\\"k-popover--base k-popover--base__top k-popover--base__top__dark k-popover--base__dark svelte-1dixidq\\" data-popper-placement=\\"\\" aria-hidden=\\"true\\" style=\\"position: fixed; left: 0px; top: 0px; margin: 0px; animation: __svelte_185617033_0 200ms linear 0ms 1 both;\\"><div slot=\\"contentEl\\" class=\\"flex flex-col\\"><p class=\\"!my-2\\">有美一人，清扬婉兮</p></div> <div k-popover-arrow=\\"true\\" data-popper-arrow-bottom=\\"\\" data-popper-arrow-top=\\"\\" data-popper-arrow-right=\\"\\" data-popper-arrow-left=\\"\\" class=\\"svelte-1dixidq\\"></div></div></div>"`;

exports[`Test: KPopover > slots: triggerEl & contentEl 1`] = `"<div><div aria-hidden=\\"true\\"><div slot=\\"triggerEl\\" cls=\\"mx-2\\">triggerEl</div></div> <div class=\\"k-popover--base k-popover--base__top k-popover--base__top__dark k-popover--base__dark svelte-1dixidq\\" data-popper-placement=\\"\\" aria-hidden=\\"true\\" style=\\"position: fixed; left: 0px; top: 0px; margin: 0px; animation: __svelte_185617033_0 200ms linear 0ms 1 both;\\"><div slot=\\"contentEl\\" class=\\"flex flex-col\\"><p class=\\"!my-2\\">contentEl</p></div> <div k-popover-arrow=\\"true\\" data-popper-arrow-bottom=\\"\\" data-popper-arrow-top=\\"\\" data-popper-arrow-right=\\"\\" data-popper-arrow-left=\\"\\" class=\\"svelte-1dixidq\\"></div></div></div>"`;
